pci/src/vfio.rs

```diff
@@ -1173,42 +1190,46 @@ impl VfioPciDevice {
                 }
 
                 let mmap_size = self.device.get_region_size(region.index);
-                let offset = self.device.get_region_offset(region.index);
-
-                let host_addr = unsafe {
-                    libc::mmap(
-                        null_mut(),
-                        mmap_size as usize,
-                        prot,
-                        libc::MAP_SHARED,
-                        fd,
-                        offset as libc::off_t,
-                    )
-                };
+                let mmap_offset = self.device.get_region_offset(region.index);
 
-                if host_addr == libc::MAP_FAILED {
-                    error!(
-                        "Could not mmap region index {}: {}",
-                        region.index,
-                        io::Error::last_os_error()
-                    );
-                    continue;
-                }
-
-                // In case the region that is being mapped contains the MSI-X
-                // vectors table or the MSI-X PBA table, we must adjust what
-                // is being declared through the hypervisor. We want to make
-                // sure we will still trap MMIO accesses to these MSI-X
-                // specific ranges.
-                let user_memory_regions = Self::generate_user_memory_regions(
+                let sparse_areas = Self::generate_sparse_areas(
+                    &caps,
                     region.index,
-                    region.start.raw_value(),
+                    region.start.0,
                     mmap_size,
-                    host_addr as u64,
-                    &mem_slot,
                     self.common.interrupt.msix.as_ref(),
-                );
-                for user_memory_region in user_memory_regions.iter() {
+                )?;
+
+                let mut user_memory_regions = Vec::new();
+                for area in sparse_areas.iter() {
+                    let host_addr = unsafe {
+                        libc::mmap(
+                            null_mut(),
+                            area.size as usize,
+                            prot,
+                            libc::MAP_SHARED,
+                            fd,
+                            mmap_offset as libc::off_t + area.offset as libc::off_t,
+                        )
+                    };
+
+                    if host_addr == libc::MAP_FAILED {
+                        error!(
+                            "Could not mmap sparse area (offset = 0x{:x}, size = 0x{:x}): {}",
+                            area.offset,
+                            area.size,
+                            std::io::Error::last_os_error()
+                        );
+                        return Err(VfioPciError::MmapArea);
```

Comment:
Before return the error, do we need to cleanup the successful mmaps that are already done from the previous iterations of the for loop?

Reply:
I don't think so. The error will cause the vmm thread to terminate, meaning the Cloud Hypervisor process will exit. Therefore all the mmaps will be cleaned by the OS.

Reply:
What about the case for hotplug with `add-device`? IIUC, the error would end up with a `HttpError::ApiError` responded to the API requester, while the vmm thread keeps running, right?

Reply:
This is a good point. Could you please open an issue for that? I can take care of it tomorrow.
Do we need to handle the same thing for vfio-user by the way?

Reply:
Good point. I thought 'vfio_user' was fine, as it did not return on mmap errors, but it did return on other errors prematurely and will be prone to a similar leak. I just opened an issue #4068 and a quick fix #4069. PTAL.